sample/sample.go

```diff
@@ -0,0 +1,205 @@
+package sample
+
+import (
+	"cmp"
+	"errors"
+	"math"
+	"slices"
+
+	"golang.org/x/exp/rand"
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/stat/sampleuv"
+)
+
+type Transform interface {
+	Apply([]float64) ([]float64, error)
+}
+
+type Sampler interface {
+	Sample([]float64) (int, error)
+}
+
+type SamplerChain struct {
+	transforms []Transform
+	sampler    Sampler
+}
+
+func NewSampler(transforms []Transform, sampler Sampler) *SamplerChain {
+	return &SamplerChain{
+		transforms: transforms,
+		sampler:    sampler,
+	}
+}
+
+// TODO(parthsareen): potentially cache softmax values
+func softmax(logits []float64) []float64 {
+	copiedLogits := make([]float64, len(logits))
+	copy(copiedLogits, logits)
+	for i := range copiedLogits {
+		copiedLogits[i] = math.Exp(copiedLogits[i])
+	}
+
+	floatSum := floats.Sum(copiedLogits)
+	floats.Scale(1.0/floatSum, copiedLogits)
+
+	return copiedLogits
+}
+
+type Temperature float64
+
+func (t Temperature) Apply(logits []float64) ([]float64, error) {
+	if t < 0 || t > 2 {
+		return nil, errors.New("temperature must be between 0 and 2")
+	}
+
+	// subtracting max logit to avoid under/overflow
+	maxLogit := floats.Max(logits)
+
+	temp := math.Max(float64(t), 1e-7)
```

Comment:
It's no obvious you're setting a floor for temperature so it's non-zero. It should be moved up to before the validation

Reply:
Will see what this should look like - it's a bit dependent on the sampler interface. If t=0 logic lives outside then should be able to get rid of this